encrypt.go

```diff
@@ -39,47 +43,100 @@ func (es *encryptStream) Write(plaintext []byte) (int, error) {
 	if ret, es.err = es.buffer.Write(plaintext); es.err != nil {
 		return 0, es.err
 	}
-	for es.buffer.Len() >= encryptionBlockSize {
-		es.err = es.encryptBlock()
+
+	// If es.buffer.Len() == encryptionBlockSize, we don't want to
+	// write it out just yet, since for V2 we need to be sure this
+	// isn't the last block.
+	for es.buffer.Len() > encryptionBlockSize {
+		es.err = es.encryptBlock(false)
 		if es.err != nil {
 			return 0, es.err
 		}
 	}
 	return ret, nil
 }
 
-func (es *encryptStream) encryptBlock() error {
-	var n int
-	var err error
-	n, err = es.buffer.Read(es.inblock[:])
-	if err != nil {
-		return err
+func makeEncryptionBlock(version Version, ciphertext []byte, authenticators []payloadAuthenticator, isFinal bool) interface{} {
+	ebV1 := encryptionBlockV1{
+		PayloadCiphertext:  ciphertext,
+		HashAuthenticators: authenticators,
+	}
+	switch version {
+	case Version1():
+		return ebV1
+	case Version2():
+		return encryptionBlockV2{
+			encryptionBlockV1: ebV1,
+			IsFinal:           isFinal,
+		}
+	default:
+		panic(ErrBadVersion{version})
 	}
-	return es.encryptBytes(es.inblock[0:n])
 }
 
-func (es *encryptStream) encryptBytes(b []byte) error {
+func checkEncryptBlockRead(version Version, isFinal bool, blockSize, plaintextLen, bufLen int) {
+	die := func() error {
+		return fmt.Errorf("invalid encryptBlock read state: version=%s, isFinal=%t, plaintextLen=%d, bufLen=%d", plaintextLen, bufLen)
+	}
+
+	// We shouldn't read more than a full block's worth.
+	if plaintextLen > blockSize {
+		die()
```

Comment:
It looks like this function is constructing an error value and then dropping it. Am I misreading?